### PR TITLE
[quant][trt] Add support for torch.addmm in TensorRT

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68229
* #68228
* #68227
* #68042
* __->__ #67537
* #68602

Summary:
This PR adds support for quantizing torch.addmm to produce a reference quantized pattern,
and also adds support in the backend_config_dict api that allows people to specify the input, weight and bias input for each input:

```
    addmm_config = {
        "pattern": torch.addmm,
        "observation_type": ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT,
        "dtype_configs": [
            weighted_op_qint8_dtype_config,
        ],
        # a map from input type to input index
        "input_type_to_index": {
            "bias": 0,
            "input": 1,
            "weight": 2,
        }
    }
```

This requires some changes in getting weight_dtype and bias_dtype in the type inference stage of prepare, which will be added in the previous PR

Test Plan:
```
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps.test_addmm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32014998](https://our.internmc.facebook.com/intern/diff/D32014998)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32014998).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32014998).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32014998).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32014998).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32014998).

File: torch/ao/quantization/fx/backend_config_dict/tensorrt.py [hunk 1]
```diff
@@ -86,6 +86,19 @@ def get_tensorrt_backend_config_dict():
             weighted_op_qint8_dtype_config,
         ]
     }
+    addmm_config = {
```
Comment (vkuzo) [inline torch/ao/quantization/fx/backend_config_dict/tensorrt.py:89]: doesn't have to be in this PR, but is there a README or a type definition of the valid keys of this dictionary, and some validation to catch common errors?  I.e. how would a user know that this new `input_type_to_index` is a valid option and what is its format.

**Review by vkuzo — APPROVED** [review]

File: torch/ao/quantization/fx/backend_config_dict/tensorrt.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config_dict/tensorrt.py:89]: yeah planning to have a complete README after this is done. also we might want to use dataclass for these configs as suggested by Dima

Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config_dict/tensorrt.py:89]: will add a TODO for now

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32014998).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32014998).

**Pull request closed without merging**